File: dat-nuget/Saro.Dat/Dat.cs
using System.Text;

namespace Saro.Dat;

public class Dat : ICloneable
{
    public string Raw { get; }
    public long Expire { get; }
    public long Cid { get; }
    internal byte[] PlainBytes { get; }
    internal byte[] SecureBytes { get; }
    internal byte[] SignatureBytes { get; }
    internal byte[] Body { get; }

    public Dat(string dat)
    {
        if (dat is null)
            throw new DatException(DatErrorCode.ConfigArgumentInvalid, "dat string is null");

        Raw = dat;

        var parts = dat.Split('.');
        if (parts.Length != 5)
            throw new DatException(DatErrorCode.TokenMalformed, "expected exactly 5 dot-separated fields");

        Expire = Field(DatErrorCode.TokenMalformed, "expire field is not a plain decimal integer",
            () => DatUtils.ParseSecondsStrict(parts[0]));

        if (Expire <= Unixtime.Now())
            throw new DatException(DatErrorCode.TokenExpired);

        Cid = Field(DatErrorCode.TokenMalformed, "cid field is not a plain hex integer",
            () => DatUtils.ParseCidStrict(parts[1]));
        PlainBytes = Field(DatErrorCode.TokenMalformed, "plain field is not base64url",
            () => DatUtils.DecodeBase64Url(parts[2]));
        SecureBytes = Field(DatErrorCode.TokenMalformed, "secure field is not base64url",
            () => DatUtils.DecodeBase64Url(parts[3]));

        if (parts[4].Length == 0)
            throw new DatException(DatErrorCode.SigMalformed, "signature field is empty");
        SignatureBytes = Field(DatErrorCode.SigMalformed, "signature field is not base64url",
            () => DatUtils.DecodeBase64Url(parts[4]));
        if (SignatureBytes.Length == 0)
            throw new DatException(DatErrorCode.SigMalformed, "signature field is empty");

        Body = Encoding.UTF8.GetBytes(dat, 0, dat.LastIndexOf('.'));
    }

    private static T Field<T>(string code, string detail, Func<T> read)
    {
        try
        {
            return read();
        }
        catch (DatException)
        {
            throw;
        }
        catch (Exception e)
        {
            throw new DatException(code, detail, e);
        }
    }

    public object Clone() => new Dat(Raw);
}

File: dat-maven/src/main/kotlin/me/saro/dat/dat/Dat.kt
package me.saro.dat.dat

import me.saro.dat.DatUtils
import me.saro.dat.Unixtime
import me.saro.dat.exception.DatErrorCode
import me.saro.dat.exception.DatException
import me.saro.dat.exception.DatResult

class Dat private constructor(
    val dat: String,
    internal val expire: ULong,
    internal val cid: ULong,
    internal val plainBytes: ByteArray,
    internal val secureBytes: ByteArray,
    internal val signatureBytes: ByteArray,
    internal val body: ByteArray
): Cloneable {
    fun getCid(): Long = cid.toLong()
    fun getExpire(): Long = expire.toLong()

    public override fun clone(): Dat {
        return Dat(dat, expire, cid, plainBytes.clone(), secureBytes.clone(), signatureBytes.clone(), body.clone())
    }

    companion object {
        @JvmStatic
        fun parse(dat: String?): DatResult<Dat> {
            if (dat.isNullOrEmpty()) {
                return DatResult.failure(DatException(DatErrorCode.TOKEN_MALFORMED, "token is empty"))
            }
            val parts = dat.split('.')
            if (parts.size != 5) {
                return DatResult.failure(
                    DatException(DatErrorCode.TOKEN_MALFORMED, "expected exactly 5 dot-separated fields")
                )
            }

            val expire = DatUtils.parseU64OrNull(parts[0])
                ?: return DatResult.failure(
                    DatException(DatErrorCode.TOKEN_MALFORMED, "expire field is not a plain decimal u64")
                )

            if (expire <= Unixtime.now().toULong()) {
                return DatResult.failure(DatException(DatErrorCode.TOKEN_EXPIRED))
            }

            val cid = DatUtils.parseU64HexOrNull(parts[1])
                ?: return DatResult.failure(
                    DatException(DatErrorCode.TOKEN_MALFORMED, "cid field is not a plain hex u64")
                )

            if (parts[4].isEmpty()) {
                return DatResult.failure(DatException(DatErrorCode.SIG_MALFORMED, "signature field is empty"))
            }

            val plainBytes: ByteArray = try {
                DatUtils.decodeBase64Url(parts[2])
            } catch (e: Exception) {
                return DatResult.failure(
                    DatException(DatErrorCode.TOKEN_MALFORMED, "plain field is not base64url", e)
                )
            }
            val secureBytes: ByteArray = try {
                DatUtils.decodeBase64Url(parts[3])
            } catch (e: Exception) {
                return DatResult.failure(
                    DatException(DatErrorCode.TOKEN_MALFORMED, "secure field is not base64url", e)
                )
            }
            val signatureBytes: ByteArray = try {
                DatUtils.decodeBase64Url(parts[4])
            } catch (e: Exception) {
                return DatResult.failure(
                    DatException(DatErrorCode.SIG_MALFORMED, "signature field is not base64url", e)
                )
            }

            val body: ByteArray = dat.substring(0, dat.lastIndexOf('.')).toByteArray()
            return DatResult.success(Dat(dat, expire, cid, plainBytes, secureBytes, signatureBytes, body))
        }
    }
}
